cpp/log/cluster_state_controller_test.cc

```diff
@@ -0,0 +1,289 @@
+  ClusterStateControllerTest()
+      : pool_(2),
+        base_(make_shared<libevent::Base>()),
+        pump_(base_),
+        etcd_(base_),
+        store_(new EtcdConsistentStore<LoggedCertificate>(&pool_, &etcd_, "",
+                                                          kNodeId1)),
+        store2_(new EtcdConsistentStore<LoggedCertificate>(&pool_, &etcd_, "",
+                                                           kNodeId2)),
+        store3_(new EtcdConsistentStore<LoggedCertificate>(&pool_, &etcd_, "",
+                                                           kNodeId3)),
+        controller_(store_.get(), &election_, 1, 1) {
+    // Set up some handy STHs
+    sth100_.set_tree_size(100);
+    sth100_.set_timestamp(100);
+    sth200_.set_tree_size(200);
+    sth200_.set_timestamp(200);
+    sth300_.set_tree_size(300);
+    sth300_.set_timestamp(300);
+    cns100_.mutable_newest_sth()->CopyFrom(sth100_);
+    cns200_.mutable_newest_sth()->CopyFrom(sth200_);
+    cns300_.mutable_newest_sth()->CopyFrom(sth300_);
+
+    ON_CALL(election_, IsMaster()).WillByDefault(Return(false));
+  }
+
+ protected:
+  ct::ClusterNodeState GetLocalState() {
+    return controller_.local_node_state_;
+  }
+
+  ct::ClusterNodeState GetNodeStateView(const string& node_id) {
+    auto it(controller_.all_node_states_.find(node_id));
+    CHECK(it != controller_.all_node_states_.end());
+    return it->second;
+  }
+
+  SignedTreeHead sth100_, sth200_, sth300_;
+  ClusterNodeState cns100_, cns200_, cns300_;
+
+  ThreadPool pool_;
+  shared_ptr<libevent::Base> base_;
+  libevent::EventPumpThread pump_;
+  FakeEtcdClient etcd_;
+  std::unique_ptr<EtcdConsistentStore<LoggedCertificate>> store_;
+  std::unique_ptr<EtcdConsistentStore<LoggedCertificate>> store2_;
+  std::unique_ptr<EtcdConsistentStore<LoggedCertificate>> store3_;
+  NiceMock<MockMasterElection> election_;
+  ClusterStateController<LoggedCertificate> controller_;
+};
+
+
+typedef class EtcdConsistentStoreTest EtcdConsistentStoreDeathTest;
+
+
+TEST_F(ClusterStateControllerTest, TestNewTreeHead) {
+  ct::SignedTreeHead sth;
+  sth.set_tree_size(234);
+  controller_.NewTreeHead(sth);
+  EXPECT_EQ(sth.DebugString(), GetLocalState().newest_sth().DebugString());
+}
+
+
+TEST_F(ClusterStateControllerTest, TestContiguousTreeSizeUpdated) {
+  const int kNewTreeSize(2345);
+  controller_.ContiguousTreeSizeUpdated(kNewTreeSize);
+  EXPECT_EQ(kNewTreeSize, GetLocalState().contiguous_tree_size());
+}
+
+
+TEST_F(ClusterStateControllerTest, TestCalculateServingSTHAt50Percent) {
+  MockMasterElection election_is_master;
+  EXPECT_CALL(election_is_master, IsMaster()).WillRepeatedly(Return(true));
+  ClusterStateController<LoggedCertificate> controller50(store_.get(),
+                                                         &election_is_master,
+                                                         1, 0.5);
+  store_->SetClusterNodeState(cns100_);
+  sleep(1);
+  util::StatusOr<SignedTreeHead> status(
```

Comment:
Not a big deal, can leave it like this, but I usually put the emphasis on the "Or<Thing>" in "StatusOr<Thing>" (this is mainly a ```SignedTreeHead```, once you've got the status check out of the way).